### PR TITLE
Adding NLP Metrics SQuADv2 and WER

### LINKED ISSUE #51: [Metrics] ROUGE
## 🚀 Feature
Implement ROUGE

### LINKED ISSUE #52: [Metrics] WER
Add WER cc @oplatek 

https://github.com/PyTorchLightning/pytorch-lightning/issues/973#issuecomment-592420431

### LINKED ISSUE #366: Squad V2 Question Answering Task metric
## 🚀 Feature

For the Question Answering Task, we would need the SQUADv2 metrics  
I would like to have it here

### PR DESCRIPTION
**Pull request by gagan3012**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #52 
Fixes #366 (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/text/wer/wer.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from typing import Any
+
+from jiwer import compute_measures
+
+from torchmetrics.metric import Metric
+
+
+class WER(Metric):
+
+    def __init__(self, concatenate_texts=False):
```
Comment (SkafteNicki) [inline torchmetrics/text/wer/wer.py:10]:
```suggestion
    def __init__(self, concatenate_texts: bool = False):
```

File: torchmetrics/text/wer/wer.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+from typing import Any
+
+from jiwer import compute_measures
+
+from torchmetrics.metric import Metric
+
+
+class WER(Metric):
+
```
Comment (SkafteNicki) [inline torchmetrics/text/wer/wer.py:9]: missing docstring

File: torchmetrics/text/wer/wer.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self, concatenate_texts=False):
+        super().__init__()
+        self.concatenate_texts = concatenate_texts
+
+    def update(self, preds: Any, target: Any) -> None:
+        self.preds = preds
+        self.target = target
```
Comment (SkafteNicki) [inline torchmetrics/text/wer/wer.py:16]: i assume that we want to accumulate over multiple sentences
```suggestion
        self.preds.append(preds)
        self.target.append(target)
```

File: torchmetrics/text/wer/wer.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+from typing import Any
+
+from jiwer import compute_measures
```
Comment (SkafteNicki) [inline torchmetrics/text/wer/wer.py:3]: there need to be added a file similar to `https://github.com/PyTorchLightning/metrics/blob/master/requirements/image.txt` that specifies the required dependencies for the new text metrics

File: torchmetrics/text/squadv2/squadv2.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+               Args:
+                   n_gram: Gram value ranged from 1 to 4 (Default 4)
+                   smooth: Whether or not to apply smoothing – Lin et al. 2004
+               """
+        super().__init__()
+        self.n_gram = n_gram
+        self.smooth = smooth
+        self.no_answer_threshold = no_answer_threshold
+
+    def update(self, predictions, references) -> None:
```
Comment (SkafteNicki) [inline torchmetrics/text/squadv2/squadv2.py:71]:
```suggestion
    def update(self, predictions: Any, references: Any) -> None:
```

File: torchmetrics/text/squadv2/squadv2.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        self.predictions = predictions
+        self.references = references
+
+    def compute(self):
```
Comment (SkafteNicki) [inline torchmetrics/text/squadv2/squadv2.py:75]:
```suggestion
    def compute(self) -> Dict[str, float]:
```

File: torchmetrics/text/squadv2/evaluate_squad.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
+import argparse
```
Comment (SkafteNicki) [inline torchmetrics/text/squadv2/evaluate_squad.py:1]: If this script is copied from somewhere else then we need to include a reference.
Also can we remove anything non-essential and rename to something else like `helper.py`?

**Review by SkafteNicki — COMMENTED** [review]
Mix of comments. Just wanted to make sure: Are these metrics meant to be accumulated over multiple batches or are they only intended to be evaluated for single batch/sample? Because if it is the latter then it is maybe better to implement them as functionals.

File: torchmetrics/text/squadv2/evaluate_squad.py (hunk 7 above)
Comment (Borda) [reply torchmetrics/text/squadv2/evaluate_squad.py:1]: moreover, can you explain why such a script is even needed?

File: torchmetrics/text/squadv2/squadv2.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    merge_eval,
+)
+
+
+class SQuADv2(Metric):
+    """
+    This metric wrap the official scoring script for version 2 of the Stanford Question
+    Answering Dataset (SQuAD).
+    Computes SQuAD v2 scores (F1 and EM).
+Args:
```
Comment (Borda) [inline torchmetrics/text/squadv2/squadv2.py:18]: the formations is strange here, you shall add left padding

File: torchmetrics/text/squadv2/squadv2.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
+from torchmetrics.metric import Metric
+
+from .evaluate_squad import (
+    apply_no_ans_threshold,
+    find_all_best_thresh,
+    get_raw_scores,
+    make_eval_dict,
+    make_qid_to_has_ans,
+    merge_eval,
+)
```
Comment (Borda) [inline torchmetrics/text/squadv2/squadv2.py:10]: these shall live in `torchmetrics/functional/text/squadv2/squadv2.py`

File: torchmetrics/text/wer/wer.py (hunk 2 above)
Comment (Borda) [reply torchmetrics/text/wer/wer.py:9]: missing typing for states

File: torchmetrics/text/wer/wer.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        if self.concatenate_texts:
+            return compute_measures(self.target, self.preds)["wer"]
+        else:
+            incorrect = 0
+            total = 0
+            for prediction, reference in zip(self.preds, self.target):
+                measures = compute_measures(reference, prediction)
+                incorrect += measures["substitutions"] + measures["deletions"] + measures["insertions"]
+                total += measures["substitutions"] + measures["deletions"] + measures["hits"]
+            return incorrect / total
```
Comment (Borda) [inline torchmetrics/text/wer/wer.py:28]:
```suggestion
        incorrect = 0
        total = 0
        for prediction, reference in zip(self.preds, self.target):
            measures = compute_measures(reference, prediction)
            incorrect += measures["substitutions"] + measures["deletions"] + measures["insertions"]
            total += measures["substitutions"] + measures["deletions"] + measures["hits"]
        return incorrect / total
```

**Review by Borda — CHANGES_REQUESTED** [review]

**Comment by gagan3012** [comment]
Hello, 
I will implement all changes by tonight (EST)

File: torchmetrics/text/wer/wer.py (hunk 3 above)
Comment (gagan3012) [reply torchmetrics/text/wer/wer.py]: Yes you are right

File: torchmetrics/text/wer/wer.py (hunk 4 above)
Comment (gagan3012) [reply torchmetrics/text/wer/wer.py:3]: Yes I have created a requirements/text.txt file

File: torchmetrics/text/squadv2/evaluate_squad.py (hunk 7 above)
Comment (gagan3012) [reply torchmetrics/text/squadv2/evaluate_squad.py]: It's the helper script that has all the required functions for running the metric

File: torchmetrics/text/wer/wer.py (hunk 2 above)
Comment (gagan3012) [reply torchmetrics/text/wer/wer.py:11]: Updated

**Comment by gagan3012** [comment]
Hello, 
I have made all the changes requested. Please take a look Thanks

**Comment by Borda** [comment]
@gagan3012 seems it is your first PR so GH requires to allow running CI by a maintainer, so pls when you see it is hanging, pls ping me in slack so we can speed it up :]

File: docs/source/references/modules.rst [hunk 11]
```diff
@@ -502,6 +502,23 @@ RetrievalNormalizedDCG
 .. autoclass:: torchmetrics.RetrievalNormalizedDCG
     :noindex:
 
+****
+Text
+****
+
+SQuADV2
+~~~~~~~~~
```
Comment (Borda) [inline docs/source/references/modules.rst:510]:
```suggestion
~~~~~~~
```

File: docs/source/references/modules.rst [hunk 12]
```diff
@@ -502,6 +502,23 @@ RetrievalNormalizedDCG
...
+
+.. autoclass:: torchmetrics.SQuADV2
+    :noindex:
+
+
+WER
+~~~~~~~~~
```
Comment (Borda) [inline docs/source/references/modules.rst:517]:
```suggestion
~~~
```

File: requirements/text.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+jiwer
```
Comment (Borda) [inline requirements/text.txt:1]: any minimal version?

File: tests/text/test_wer.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+
+@pytest.mark.parametrize(
+    "hyp,ref",
+    [("hello world", "Firwww")],
+)
+def test_wer_different(hyp, ref):
+    metric = WER()
+    metric.update(hyp, ref)
+    assert metric.compute() == 1.0
```
Comment (Borda) [inline tests/text/test_wer.py:25]:
```suggestion
@pytest.mark.parametrize(
    "hyp,ref,score",
    [
        ("hello world", "hello world", 0.0),
        ("hello world", "Firwww", 1.0)
    ],
)
def test_wer_same(hyp, ref, score):
    metric = WER()
    metric.update(hyp, ref)
    assert metric.compute() == score
```

File: torchmetrics/functional/text/__init__.py [hunk 15]
```diff
@@ -0,0 +1,6 @@
+from torchmetrics.functional.text.squadv2 import apply_no_ans_threshold  # noqa: F401
+from torchmetrics.functional.text.squadv2 import find_all_best_thresh  # noqa: F401
+from torchmetrics.functional.text.squadv2 import get_raw_scores  # noqa: F401
+from torchmetrics.functional.text.squadv2 import make_eval_dict  # noqa: F401
+from torchmetrics.functional.text.squadv2 import make_qid_to_has_ans  # noqa: F401
+from torchmetrics.functional.text.squadv2 import merge_eval  # noqa: F401; noqa: F401
```
Comment (Borda) [inline torchmetrics/functional/text/__init__.py:6]:
```suggestion
```
there is no need to have then here

File: torchmetrics/functional/text/squadv2.py [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+        help='Predict "" if no-answer probability exceeds this (default = 1.0).',
+    )
+    parser.add_argument(
+        "--out-image-dir", "-p", metavar="out_images", default=None, help="Save precision-recall curves to directory."
+    )
+    parser.add_argument("--verbose", "-v", action="store_true")
+    if len(sys.argv) == 1:
+        parser.print_help()
+        sys.exit(1)
+    return parser.parse_args()
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:39]:
```suggestion
```
TM is not using CLI yet, unless there is a special reason for it...

File: torchmetrics/functional/text/squadv2.py [hunk 17]
```diff
@@ -0,0 +1,316 @@
+import argparse
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:1]: can you confirm that this is your own code? :rabbit:

File: torchmetrics/functional/text/squadv2.py [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+        ])
+
+
+def merge_eval(main_eval, new_eval, prefix) -> None:
+    for k in new_eval:
+        main_eval["%s_%s" % (prefix, k)] = new_eval[k]
+
+
+def plot_pr_curve(precisions, recalls, out_image, title) -> None:
+    plt.step(recalls, precisions, color="b", alpha=0.2, where="post")
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:152]: use fig as an anchor for plotting, this is very sensitive to actually active figure

File: torchmetrics/functional/text/squadv2.py [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+    plt.savefig(out_image)
+    plt.clf()
+
+
+def make_precision_recall_eval(scores,
+                               na_probs,
+                               num_true_pos,
+                               qid_to_has_ans,
+                               out_image=None,
+                               title=None) -> Dict[str, float]:
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:168]:
```suggestion
                               title=None,) -> Dict[str, float]:
```

File: torchmetrics/functional/text/squadv2.py [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+    merge_eval(main_eval, pr_f1, "pr_f1")
+    merge_eval(main_eval, pr_oracle, "pr_oracle")
+
+
+def histogram_na_prob(na_probs, qid_list, image_dir, name) -> None:
+    if not qid_list:
+        return
+    x = [na_probs[k] for k in qid_list]
+    weights = np.ones_like(x) / float(len(x))
+    plt.hist(x, weights=weights, bins=20, range=(0.0, 1.0))
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:232]: the same with the figure, but anyway any figure is needed?

File: torchmetrics/functional/text/squadv2.py [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+    for i, qid in enumerate(qid_list):
+        if qid not in scores:
+            continue
+        if qid_to_has_ans[qid]:
+            diff = scores[qid]
+        else:
+            if preds[qid]:
+                diff = -1
+            else:
+                diff = 0
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:255]:
```suggestion
            diff = -1 if preds[qid] else 0
```

File: torchmetrics/functional/text/squadv2.py [hunk 22]
```diff
@@ -0,0 +1,316 @@
...
+
+
+if __name__ == "__main__":
+    OPTS = parse_args()
+    if OPTS.out_image_dir:
+        import matplotlib
+
+        matplotlib.use("Agg")
+        import matplotlib.pyplot as plt
+    main()
```
Comment (Borda) [inline torchmetrics/functional/text/squadv2.py:316]: we do not want to add any entry point to TM atm the moment, correct?
cc: @PyTorchLightning/core-metrics @PyTorchLightning/core-contributors

File: torchmetrics/text/squadv2/squadv2.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+from torchmetrics.metric import Metric
+
+
+class SQuADv2(Metric):
+    """
+    This metric wrap the official scoring script for version 2 of the Stanford Question
+    Answering Dataset (SQuAD).
+    Computes SQuAD v2 scores (F1 and EM).
+    Args:
```
Comment (Borda) [inline torchmetrics/text/squadv2/squadv2.py:19]:
```suggestion

    Args:
```

File: torchmetrics/text/squadv2/squadv2.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+        predictions: List of triple for question-answers to score with the following elements:
+            - the question-answer 'id' field as given in the references (see below)
+            - the text of the answer
+            - the probability that the question has no answer
+        references: List of question-answers dictionaries with the following key-values:
+            - 'id': id of the question-answer pair (see above),
+            - 'answers': a list of Dict {'text': text of the answer as a string}
+        no_answer_threshold: float Probability threshold to decide that a question has no answer.
+    Returns:
```
Comment (Borda) [inline torchmetrics/text/squadv2/squadv2.py:28]:
```suggestion

    Returns:
```

File: torchmetrics/text/squadv2/squadv2.py [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+        'HasAns_f1': The F-score of predicted tokens versus the gold answer
+        'HasAns_total': Number of score considered
+        'NoAns_exact': Exact match (the normalized answer exactly match the gold answer)
+        'NoAns_f1': The F-score of predicted tokens versus the gold answer
+        'NoAns_total': Number of score considered
+        'best_exact': Best exact match (with varying threshold)
+        'best_exact_thresh': No-answer probability threshold associated to the best exact match
+        'best_f1': Best F1 (with varying threshold)
+        'best_f1_thresh': No-answer probability threshold associated to the best F1
+    Examples:
```
Comment (Borda) [inline torchmetrics/text/squadv2/squadv2.py:42]:
```suggestion

    Examples:
```

File: torchmetrics/text/squadv2/squadv2.py [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+        'best_f1': 100.0,
+        'best_f1_thresh': 0.0}
+    """
+
+    def __init__(self, n_gram: int = 4, smooth: bool = False, no_answer_threshold: float = 1.0):
+        """
+               Args:
+                   n_gram: Gram value ranged from 1 to 4 (Default 4)
+                   smooth: Whether or not to apply smoothing – Lin et al. 2004
+               """
```
Comment (Borda) [inline torchmetrics/text/squadv2/squadv2.py:65]:
```suggestion
       Args:
           n_gram: Gram value ranged from 1 to 4 (Default 4)
           smooth: Whether or not to apply smoothing – Lin et al. 2004
       """
```

File: torchmetrics/text/wer/wer.py [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    """
+    Word error rate (WER) is a common metric of the performance of an automatic speech recognition system.
+
+    WER's output is always a number between 0 and 1.
+    This value indicates the percentage of words that were incorrectly predicted.
+    The lower the value, the better the performance of the ASR system with a WER of 0 being a perfect score.
+
+    Word error rate can then be computed as:
+    WER = (S + D + I) / N = (S + D + I) / (S + D + C)
+    where
```
Comment (Borda) [inline torchmetrics/text/wer/wer.py:18]:
```suggestion
    where:
```

File: torchmetrics/text/wer/wer.py [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+    C is the number of correct words,
+    N is the number of words in the reference (N=S+D+C).
+
+    Compute WER score of transcribed segments against references.
+
+    Args:
+        references: List of references for each speech input.
+        predictions: List of transcriptions to score.
+        concatenate_texts (bool, default=False): Whether to concatenate all input texts or compute WER iteratively.
+    Returns:
```
Comment (Borda) [inline torchmetrics/text/wer/wer.py:31]:
```suggestion

    Returns:
```

File: torchmetrics/text/wer/wer.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+        (float): the word error rate
+    Examples:
```
Comment (Borda) [inline torchmetrics/text/wer/wer.py:33]:
```suggestion

    Examples:
```

**Review by Borda — CHANGES_REQUESTED** [review]

**Review by Borda — COMMENTED** [review]
I am not very convinced about having any entry point to TM, that shall be done consistently to all metrics then...
also, WEB testing is very sparse and missing any tests for squadv2

@gagan3012 we really appreciate your effort, I would just recommend splitting adding the two metrics to two PR where each will be adding only one of them which would simplify review and also your work by focussing only on one at the time :rabbit: 

cc: @SkafteNicki

**Comment by gagan3012** [comment]
Hello, 
I think I will close this and Create separate PRs for both of the metrics

**Pull request closed without merging**